### PR TITLE
Include pandoc binaries

### PR DESCRIPTION
**Pull request by JanSchulz**
This is a first version how to include the pandoc binaries. Only windows is currently implemented...

```
python setup.py download_pandoc
python setup.py bdist_wheel
```

Will first download the pandoc binaries and put them in `pypandoc/files`. The bdist_wheel step then includes the files in the wheel package. Pandoc now searches for pandoc first on the path and then on the new location in the package and takes the bigger pandoc version (equal will use the one on path). 

I move the `pypandoc.py` file to `pypandoc.__init__.py` as I wanted to have all files in a subdir and not directly in site_packages.

replaces: #37 
Closes: #16 
Closes:: #72: env var for path to pandoc

### CODE REVIEW

**Comment by JanSchulz** [comment]
If this is taken (aka the move from pypandoc.py to pypandoc/__init__.py, then I think it makes sense to split out the py3compat stuff to a different file...

**Comment by bebraw** [comment]
> If this is taken (aka the move from pypandoc.py to pypandoc/init.py, then I think it makes sense to split out the py3compat stuff to a different file...

I'm not exactly sure how people like to deal with this in Python but that sounds reasonable.

**Comment by JanSchulz** [comment]
Yay, it's green...

**Comment by JanSchulz** [comment]
Ok, this has one big drawback right now: it builds a wheel with 'any' in the filename, so no windows/darwin specific wheel... Lets see...

This includes now the split of py3compat stuff into a new file. If someone with a macosx system wants to try the macosx part or somone the linux part, I'd love to include that...

**Comment by JanSchulz** [comment]
Ok, now we get platform specific wheels, but the egg seems to be still platform agnostic. Not sure what to do with it, I've never used eggs...

**Comment by JanSchulz** [comment]
Fixed the version number screwup and also fixed some commits which mixed wrong stuff...

**Comment by bebraw** [comment]
Do you expect that the binary distributions will be available through PyPI? I don't mind merging your work. Just curious.

**Comment by JanSchulz** [comment]
I would be very happy if there would be binary wheels on PyPi.

I see the problem that pandoc is GPL, so that means that I should add the license to some place of the package and theoretically someone could then ask you to send a physical cd with the sources...

http://stackoverflow.com/questions/5437501/can-i-use-gpl-software-binaries-in-commercial-environment

File: setup.py [hunk 1]
```diff
@@ -9,6 +22,111 @@
     # pandoc is not installed, fallback to using raw contents
     long_description = open('README.md').read()
 
+# Uses sys.platform keys, but removes the 2 from linux2
+# Adding a new plattform means implemnting unpacking in "DownloadPandocCommand" and adding the URL here
```
Comment (bebraw) [inline setup.py]: typo: implemnting -> implementing

File: setup.py [hunk 2]
```diff
@@ -9,6 +22,111 @@
...
+    "win32": "https://github.com/jgm/pandoc/releases/download/1.15.1.1/pandoc-1.15.1.1-windows.msi",
+    #"linux": "https://github.com/jgm/pandoc/releases/download/1.15.1/pandoc-1.15.1-1-amd64.deb",
+    #"darwin": "https://github.com/jgm/pandoc/releases/download/1.15.1/pandoc-1.15.1-osx.pkg"
+}
+
+class DownloadPandocCommand(Command):
+
+    """Download pandoc"""
+
+    description = "downloads a pandc release and adds it to the package"
```
Comment (bebraw) [inline setup.py]: pandc -> pandoc

File: setup.py [hunk 3]
```diff
@@ -9,6 +22,111 @@
...
+    def finalize_options(self):
+         pass
+
+    def _unpack_win32(self, filename):
+        targetfolder = os.path.join(os.path.dirname(os.path.realpath(__file__)), "pypandoc", "files")
+        print("* Unpacking %s to tempfolder..." % (filename))
+
+        tempfolder = tempfile.mkdtemp()
+
+        # Make sure the targetfolder exists...
```
Comment (bebraw) [inline setup.py]: targetfolder -> target folder perhaps

**Comment by bebraw** [comment]
> I would be very happy if there would be binary wheels on PyPi.

How do other packages deal with binaries? What's the standard?

**Comment by JanSchulz** [comment]
> How do other packages deal with binaries? What's the standard?

Actually I haven't seen any packages which do/need this.

IMO this is the right approach, at least when staying in the python world (e.g. distributor should not use included binaries and maybe conda should work by building a pandoc package which puts pandoc in PATH (script dir?). I'm not really happy with the wheel hack, but I currently see no alternatives (hm, maybe monkeypatching `self.distribution.has_ext_...()` whatever distributin is in that context).

**Comment by bebraw** [comment]
Do you know how big the binaries are? Do you need some specific environment to upload them? I'm on OS X myself btw. I can likely test that Darwin part if needed.

**Comment by JanSchulz** [comment]
The two static files on windows (`pandoc.exe` and `pandoc_citeproc.exe`) are ~80MB... The `pypandoc-1.0.3-cp27-none-win_amd64.whl` is ~20MB

The package only includes the prebuilt native binaries and otherwise only python code, so no special built env is needed, but you do currently need a OS for each supported plattform (so windows, macosx and it's probably ok to assume that linux users can take care for themself?)

Upload (using [twine](https://pypi.python.org/pypi/twine/)) is then 
```
twine upload dist/*
```

I can build windows binaries and send them to you via dropbox...

There is also https://github.com/conda-forge/conda-smithy which can hopefully gain some other PyPi uploads, too...

A completely different way would be to build the packages by adding a target platform (`python setup.py download_pandoc windows`) and then renaming the built wheel. That should work, as far as I know the only target specific thing is the name of the wheel.

So a `release.sh`:

```
python setup.py download_pandoc windows
python setup.py bdist_wheel
# rename wheel
# remove prebuilt binaries
python setup.py download_pandoc linux
...
```

**Comment by JanSchulz** [comment]
Ok, I managed to setup a smithy for pypandoc (but I will expand that to other packages) which uploads conda packages to my channel (I need pypandoc for my knitpy project).

https://github.com/JanSchulz/package-builder/

I think adding uploads to pypi will be possible by extending the [build scripts](https://github.com/JanSchulz/package-builder/tree/master/recipes/pypandoc) with some additional commands:

* download the binaries
* build the wheel
* upload to PyPi
* cleanup
* install for conda building...

I currently don't need it for knitpy (I will only build one wheel, which I will let the normal travis build upload), but I can setup the thing for pypi and you can fork and change the tokens for your accounts?

**Comment by bebraw** [comment]
@JanSchulz What if I upgraded you into a maintainer status? I'm absolutely certain you would be great at that role given your contributions so far. That way you could publish versions whenever you need and so on. I can of course remain in an advisory role but given I don't do any Python development beyond this project, I'm a little out of the loop so to speak.

**Comment by JanSchulz** [comment]
@bebraw I would be happy to be added as a maintainer! :-) 

I have to say that I'm happy with pypandoc as it is (minus the not included binaries): it does exactly what I need and I currently wouldn't do much development on pypandoc if there is no pressing issue/request coming up.

**Comment by bebraw** [comment]
@JanSchulz Alright. I added you to the GitHub project collaborators so you should have access to the repository. You are also an owner of the PyPI package now so you can release as you wish.

I'm not expecting any wonders. Just responding to the issues every once in a while will do just fine. If you have time/interest at some point, there are a couple of older issues to study.

**Comment by JanSchulz** [comment]
Thank you! I will try to setup the wheel building. 

If you want to merge this before and make a release, please go ahead.

**Comment by bebraw** [comment]
@JanSchulz I published `1.1.0` to signify the start of the new series. There's just a source dist for now. Feel free to proceed with the binaries as you want.

**Comment by JanSchulz** [comment]
@bebraw Could you add a `_unpack_darwin` function to setup.py? That's what currently missing to push it...

http://stackoverflow.com/questions/11298855/how-to-unpack-and-pack-pkg-file

Should be mostly c&p from here or if that's possible to directly unpack into the right dir...
```
    def _unpack_win32(self, filename):
        targetfolder = os.path.join(os.path.dirname(os.path.realpath(__file__)), "pypandoc", "files")
        print("* Unpacking %s to tempfolder..." % (filename))

        tempfolder = tempfile.mkdtemp()

        # Make sure targetfolder exists...
        try:
            os.makedirs(targetfolder)
        except OSError:
            pass # dir already exists...

        cmd = ["msiexec", "/a", filename, "/qb", "TARGETDIR=%s" % (tempfolder)]
        # if only 3.5 is supported, should be `run(..., check=True)`
        subprocess.check_call(cmd)

        # pandoc.exe, pandoc-citeproc.exe, and the COPYRIGHT are in the Pandoc subfolder
        for exe in ["pandoc.exe", "pandoc-citeproc.exe", "COPYRIGHT.txt"]:
            src = os.path.join(tempfolder, "Pandoc", exe)
            dst = os.path.join(targetfolder, exe)
            print("* Copying %s to %s ..." % (exe, targetfolder))
            print("%s, %s" % (src, dst))
            shutil.copyfile(src, dst)

        # remove temporary dir
        shutil.rmtree(tempfolder)
        print("* Done.")
```

**Comment by bebraw** [comment]
@JanSchulz I sketched out something. I ran into a problem, though:

```bash
pypandoc * [master] $ python setup.py download_pandoc
running download_pandoc
* Downloading pandoc from https://github.com/jgm/pandoc/releases/download/1.15.1/pandoc-1.15.1-osx.pkg ...
Traceback (most recent call last):
  File "setup.py", line 184, in <module>
    cmdclass=cmd_classes
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "setup.py", line 126, in run
    with urlopen(url) as response:
AttributeError: addinfourl instance has no attribute '__exit__'
```

Any idea what might cause that?

**Comment by JanSchulz** [comment]
Seems that urlopen on py27 isn't implemented as a context manager :-(

http://stackoverflow.com/questions/30627937/tracebaclk-attributeerroraddinfourl-instance-has-no-attribute-exit


```
        with urlopen(url) as response:
                with open(filename, 'wb') as out_file:
                    shutil.copyfileobj(response, out_file)
```

-> 

```
        response = urlopen(url)
        with open(filename, 'wb') as out_file:
            shutil.copyfileobj(response, out_file)
```

**Comment by JanSchulz** [comment]
If you are at it: in setup.py the checked exception needs to change from `OSError` to  `RuntimeError`...
-> https://travis-ci.org/JanSchulz/package-builder/builds/86807278

Damn :-(

**Comment by bebraw** [comment]
> If you are at it: in setup.py the checked exception needs to change from OSError to RuntimeError...
-> https://travis-ci.org/JanSchulz/package-builder/builds/86807278

Done.

Getting close with Darwin.

**Comment by JanSchulz** [comment]
Wow, that was fast! :-)

this line can probably also go, it's a leftover from my debugging session: `print("%s, %s" % (src, dst))`

**Comment by JanSchulz** [comment]
Ok, the assert for 64 bit can also go, at least on windows:

```
λ file ..\pypandoc\pypandoc\files\pandoc.exe
..\pypandoc\pypandoc\files\pandoc.exe: PE32 executable for MS Windows (console) Intel 80386 32-bit
```

-> So it will run fine on 32 bit. I suspect that's the same for OSX...

**Comment by bebraw** [comment]
Ok, I did those tweaks. `python setup.py download_pandoc` appears to work for me now.

**Comment by JanSchulz** [comment]
Yay, and there come the wheels:
https://testpypi.python.org/pypi/pypandoc (right now only windows yet...)
https://anaconda.org/janschulz/pypandoc/files (deleted most of them again, waiting for the macosx build to succeed...)

**Comment by bebraw** [comment]
Ok, I have no idea how Linux would work out but if someone needs that, I guess that person will implement the portion.

**Comment by JanSchulz** [comment]
Linux would be nice, as that would give someone using our wheels/conda packages the ability to use them on travis without working around older pandoc versions...

It's actually this two commands:
```
ar x xxx.deb
tar xzf data.tar.gz
```

(and to figure out how to make `ar` do it in a temp directory...)

**Pull request merged**